Mathlib/Data/Num/Lemmas.lean

```diff
@@ -0,0 +1,1819 @@
+then trying to call `simp`.
+```lean
+example (n : ZNum) (m : ZNum) : n ≤ n + m * m := by
+  transfer
+  exact mul_self_nonneg _
+```
+-/
+scoped macro (name := transfer) "transfer" : tactic => `(tactic|
+    (intros; transfer_rw; try simp [add_comm, add_left_comm, mul_comm, mul_left_comm]))
+
+instance linearOrder : LinearOrder ZNum where
+  lt := (· < ·)
+  lt_iff_le_not_le := by
+    intro a b
+    transfer_rw
+    apply lt_iff_le_not_le
+  le := (· ≤ ·)
+  le_refl := by transfer
+  le_trans := by
+    intro a b c
+    transfer_rw
+    apply le_trans
+  le_antisymm := by
+    intro a b
+    transfer_rw
+    apply le_antisymm
+  le_total := by
+    intro a b
+    transfer_rw
+    apply le_total
+  decidable_eq := instDecidableEqZNum
+  decidable_le := ZNum.decidableLE
+  decidable_lt := ZNum.decidableLT
+#align znum.linear_order ZNum.linearOrder
+
+instance addCommGroup : AddCommGroup ZNum where
+  add := (· + ·)
+  add_assoc := by transfer
+  zero := 0
+  zero_add := zero_add
+  add_zero := add_zero
+  add_comm := by transfer
+  neg := Neg.neg
+  add_left_neg := by transfer
+#align znum.add_comm_group ZNum.addCommGroup
+
+instance addMonoidWithOne : AddMonoidWithOne ZNum :=
+  { ZNum.addCommGroup with
+    one := 1
+    natCast := fun n => ZNum.ofInt' n
+    natCast_zero := show (Num.ofNat' 0).toZNum = 0 by rw [Num.ofNat'_zero]; rfl
+    natCast_succ := fun n =>
+      show (Num.ofNat' (n + 1)).toZNum = (Num.ofNat' n).toZNum + 1 by
+        rw [Num.ofNat'_succ, Num.add_one, Num.toZNum_succ, ZNum.add_one] }
+#align znum.add_monoid_with_one ZNum.addMonoidWithOne
+
+-- Porting note: These theorems should be declared out of the instance, otherwise timeouts.
+
+private theorem mul_comm : ∀ (a b : ZNum), a * b = b * a := by transfer
+
+private theorem add_le_add_left : ∀ (a b : ZNum), a ≤ b → ∀ (c : ZNum), c + a ≤ c + b := by
+  intro a b h c
+  revert h
+  transfer_rw
+  exact fun h => _root_.add_le_add_left h c
+
+instance linearOrderedCommRing : LinearOrderedCommRing ZNum :=
+  { ZNum.linearOrder, ZNum.addCommGroup, ZNum.addMonoidWithOne with
+    mul := (· * ·)
+    mul_assoc := by transfer
+    zero_mul := by transfer
+    mul_zero := by transfer
+    one := 1
+    one_mul := by transfer
+    mul_one := by transfer
+    left_distrib := by
+      transfer
+      simp [mul_add]
+    right_distrib := by
+      transfer
+      simp [mul_add, _root_.mul_comm]
+    mul_comm := mul_comm
+    exists_pair_ne := ⟨0, 1, by decide⟩
+    add_le_add_left := add_le_add_left
+    mul_pos := fun a b =>
+      show 0 < a → 0 < b → 0 < a * b by
+        transfer_rw
+        apply mul_pos
+    zero_le_one := by decide }
+#align znum.linear_ordered_comm_ring ZNum.linearOrderedCommRing
+
+@[simp, norm_cast]
+theorem cast_sub [Ring α] (m n) : ((m - n : ZNum) : α) = m - n := by simp [sub_eq_neg_add]
+#align znum.cast_sub ZNum.cast_sub
+
+-- Porting note: Required for the notation `-[n+1]`.
+open Int in
```

Comment:
Can you just open `Int` for the whole file, or section, or something? I don't think this needs repeating so many times.